python/ray/data/_internal/execution/operators/limit_operator.py

```diff
@@ -0,0 +1,86 @@
+from typing import Deque, Optional
+from collections import deque
+from ray.data._internal.stats import StatsDict
+from ray.data._internal.execution.interfaces import (
+    PhysicalOperator,
+    RefBundle,
+)
+from ray.data._internal.split import _split_at_indices
+
+
+class LimitOperator(PhysicalOperator):
+    """Physical operator for limit."""
+
+    def __init__(
+        self,
+        limit: int,
+        input_op: PhysicalOperator,
+    ):
+        self._limit = limit
+        self._consumed_rows = 0
+        self._buffer: Deque[RefBundle] = deque()
+        self._stats: StatsDict = {}
+        self._num_outputs_total = input_op.num_outputs_total()
+        if self._num_outputs_total is not None:
+            self._num_outputs_total = min(self._num_outputs_total, limit)
+
+        name = f"LimitOperator[limit={limit}]"
+        super().__init__(name, [input_op])
+
+    def _limit_reached(self) -> bool:
+        return self._consumed_rows >= self._limit
+
+    def add_input(self, refs: RefBundle, input_index: int) -> None:
+        assert not self.completed()
+        assert input_index == 0, input_index
+        if self._limit_reached():
+            return
+        input_rows = refs.num_rows()
+        if input_rows is None:
+            # If we don't know the number of rows in the input, try to
+            # split at the maximum number of rows we can consume
+            # (`self._limit - self._consumed_rows`).
+            blocks_splits, metadata_splits = _split_at_indices(
```

Comment:
Hmm this is a blocking call right? I think we want to ensure all operators are streaming compatible. Maybe there should be an op that we insert before the Limit that ensures all bundles have num_rows set, fetching it if needed.

Reply:
Actually, can we assume num_rows is always available? After the read, I think this should always be populated.